File: 2685-first-completely-painted-row-or-column/2685-first-completely-painted-row-or-column.py
class Solution:
    def firstCompleteIndex(self, arr: List[int], mat: List[List[int]]) -> int:
        def display(mat):
            for row in mat:
                print(row)
            print('-----------')
        row,col=len(mat),len(mat[0])
        hash=dict()
        for i in range(row):
            for j in range(col):
                hash[mat[i][j]]=(i,j)
        #print('Initial hash:',hash)
        #display(mat)
        rowHash={i:0 for i in range(row)}
        colHash={i:0 for i in range(col)}
        #print(rowHash,colHash)
        for i in range(len(arr)):
            x,y=hash[arr[i]]
            rowHash[x]+=1
            colHash[y]+=1
            #print(x,y,i,rowHash,colHash,rowHash[x],colHash[y])
            if rowHash[x]==col or colHash[y]==row:
                return i
        return -1

File: 2685-first-completely-painted-row-or-column/2685-first-completely-painted-row-or-column.java
class Solution {
    public int firstCompleteIndex(int[] arr, int[][] mat) {
        int row=mat.length;
        int col=mat[0].length;
        Map<Integer,int[]> hash=new HashMap<>();
        //populating hahs to access O(1)!
        for(int i=0;i<row;i++){
            for(int j=0;j<col;j++){
                hash.put(mat[i][j],new int[]{i,j});
            }
        }
        //we can use array instead of dictionary to calculate freq!!
        int[] rowHash=new int[row];
        int[] colHash=new int[col];
        //key-logic
        for(int i=0;i<arr.length;i++){
            int[] pop=hash.get(arr[i]);
            int x=pop[0];
            int y=pop[1];
            rowHash[x]++;
            colHash[y]++;
            if(rowHash[x]==col || colHash[y]==row){
                return i;
            }
        }
        return -1;


    }
}